### PR TITLE
Divide card content into left and right containers

### LINKED ISSUE #266: Split card content into left and right containers
**Description**
After implementing the 'Responsive Layout #259' feature, our application should support a new layout with two distinct containers: one on the left side for general views and one on the right side for detailed information

**Describe the solution you'd like**
The solution involves splitting the content of each step into two containers once the 'Responsive Layout' feature is in place. The left container will display general instructions or options related to the current step, while the right container will provide detailed views and interactions based on the selections made in the left container.

For example, in the first step titled "Choose Project," the left container should display two buttons: "Choose project" and "Input project manually." By default, the "Choose project" button will be selected. The right container will then show a detailed view for selecting a project to create a DMP. If the user switches to "Input project manually," the right container will update to show the relevant form fields for manual input.

**Describe alternatives you've considered**
The layout should be responsive, ensuring that on smaller screens

**Additional context**
Screenshot 2024-07-22 at 12 45 39

_If this issue requests multiple changes consider using checkboxes or lists._

### LINKED ISSUE #352: update responsiveness: step 1 & 2
**Changes**
Make the app fully functional on large and medium screens. This includes adding clear titles, appropriate spacing, and layout adjustments to make components easy to read or access.

**Additional notes**
Create a consistent user experience across screen sizes by adjusting layout and visibility without compromising functionality. 

**Assumptions**
List any assumptions that you are aware of but still need to be reviewed.

**Acceptance Criteria**
Titles and spacing adjustments will enhance readability without affecting functionality.
All features should remain fully accessible on large screens  and medium screens (laptops, iPads)

**Risks? (optional)**
Potential risks and how to best mitigate them.

### PR DESCRIPTION
**Pull request by lpandath**
## Description
This PR introduces an instruction toggle with buttons located on the left side of the interface, allowing users to navigate between the primary and secondary views of the content. The left side includes the step-by-step navigation, while the right side dynamically displays the relevant content based on the selected step and view.

#### What does this PR do?

Screenshot 2024-10-07 at 17 15 01

### Checks

- [ ] Summary updated
- [ ] Version view updated
- [ ] Documentation added
- [ ] Tests added
- [ ] E2e tests created
- [ ] Successfully ran e2e tests before merge

closes GH-

### CODE REVIEW

File: apps/damap-frontend/src/app/components/layout/layout.component.html [hunk 1]
```diff
@@ -126,7 +126,7 @@
     </div>
   </mat-sidenav>
   <mat-sidenav-content class="mat-sidenav-content">
-    <app-env-banner *ngIf="env === 'DEV'"></app-env-banner>
+    <!-- <app-env-banner *ngIf="env === 'DEV'"></app-env-banner> -->
```
Comment (ValentinFutterer) [inline apps/damap-frontend/src/app/components/layout/layout.component.html]: Could you please uncomment this, we need the banner for our demo version 😄

**Comment by ValentinFutterer** [comment]
It looks very good, i have just a few small general things i would like to add:

To switch between the views, i would suggest to use mat-tab-group instead of using buttons.
image

I could also see the use in a button on the end of each component, that jumps to the next stepper. If one component gets very long, the user would have to scroll up. Or in the case of step 9, down. 

Maybe it would improve the look if the stepper and content component would always have the same size, so that they end at the same point. Could just be me though :)
image

On smaller screens, content gets cut, off, e.g.g this happens on my work laptop, which is relatively big:
On even smaller screens and befoire the switch to mobile, the cutoff gets progressively worse. This should be addressed somehow. Maybe by making the stepper component slimmer and adding horizontal scrolling to the main content component  and automatically close the sidebar to save on space.
image

On mobile, the cutoff also happens, but since we dont try to support mobile perfectly i think thats fine for now, as long as users can still scroll around on touchscreen. I cant confirm that though. If they cannot, then its basically unusable on mobile, since even the action bar is cut off.
image

**Review by rekt-hard — COMMENTED** [review]
**General**
This is a huge feature. Thank you for digging into this!
Observed issues are listed below.

Moving forward, maybe it makes sense to split this feature into smaller chunks (basically as you already did with separate commits)
- have a feature branch (this one)
- for each step
  - create an issue
  - align on the design
    - which options should be on the left side?
    - what should be possible on the right side?
    - general information that should be displayed on the right side for either option (e.g. project for step one)
- work on each step/issue separately
- create PR for feature branch
 
When all separate issues for the split view are fixed and merged, commits of the feature branch can be squashed (step is working as expected).
Rebase onto next if required (probably a good idea to keep up-to-date with bug fixes and changes)
Create a PR (or keep this one open as a draft), merging the feature branch into next.

Even though I am not a huge fan of long lasting feature branches, this is probably a valid approach here. And a change like this should only get in if we are certain everything is still usable as it was before.

Otherwise unexpected layout behaviour or illogical UX might get into the main branch. When doing a release in-between, the commits would have to be removed manually - which is quite tedious with larger changes.

**Issues**
Components have been implemented to fit the previous design. Some behave in an illogical way now or do not fit the new layout. This is probably expected when changing the overall layout without adapting each component. As explained before, getting this feature in should ensure that everything is usable in a proper way.

1) Probably makes sense to have the stepper fixed/sticky in order to be always visible
image

2) Have a fixed/sticky footer
image

3) In step 3, the right content will wrap depending on its size when selecting an option on the left side (example width of 1120px)
image

image

4) In step 3, overflowing content will be hidden without a possibility to scroll (1300px < width < 1360px). It is fine when the menu is closed.
image
image

5) Is the gap between steps on purpose when there are no options for the left side? Maybe it makes sense to highlight the selected step some other way? Maybe use an outline icon style for other steps than the current one? This was probably not an issue before as it was clear which step a user is currently on.
image

6) Step 5 has cut off content
image

7) Step 7. Marking a dataset as deleted wraps the element
image
image


8) On smaller screens, the vertical stepper takes up quite some space and it becomes tedious having to scroll up and down. This happens when trying to change the option for the same step or when trying to get to the next step every time. Maybe it makes sense to have `next`/`previous` buttons somewhere? Or have the stepper horizontally instead of vertically for smaller screens? 
 
(Screenshots from a width of 930px which is probably not that small after all)
image
image

**Comment by lpandath** [comment]
> **General** This is a huge feature. Thank you for digging into this! Observed issues are listed below.
> 
> Moving forward, maybe it makes sense to split this feature into smaller chunks (basically as you already did with separate commits)
> 
>     * have a feature branch (this one)
> 
>     * for each step
>       
>       * create an issue
>       * align on the design
>         
>         * which options should be on the left side?
>         * what should be possible on the right side?
>         * general information that should be displayed on the right side for either option (e.g. project for step one)
> 
>     * work on each step/issue separately
> 
>     * create PR for feature branch
> 
> 
> When all separate issues for the split view are fixed and merged, commits of the feature branch can be squashed (step is working as expected). Rebase onto next if required (probably a good idea to keep up-to-date with bug fixes and changes) Create a PR (or keep this one open as a draft), merging the feature branch into next.
> 
> Even though I am not a huge fan of long lasting feature branches, this is probably a valid approach here. And a change like this should only get in if we are certain everything is still usable as it was before.
> 
> Otherwise unexpected layout behaviour or illogical UX might get into the main branch. When doing a release in-between, the commits would have to be removed manually - which is quite tedious with larger changes.
> 
> **Issues** Components have been implemented to fit the previous design. Some behave in an illogical way now or do not fit the new layout. This is probably expected when changing the overall layout without adapting each component. As explained before, getting this feature in should ensure that everything is usable in a proper way.
> 
>     1. Probably makes sense to have the stepper fixed/sticky in order to be always visible
>        image
> 
>     2. Have a fixed/sticky footer
>        image
> 
>     3. In step 3, the right content will wrap depending on its size when selecting an option on the left side (example width of 1120px)
>        image
> 
> 
> image
> 
>     4. In step 3, overflowing content will be hidden without a possibility to scroll (1300px < width < 1360px). It is fine when the menu is closed.
>        image
>        image
> 
>     5. Is the gap between steps on purpose when there are no options for the left side? Maybe it makes sense to highlight the selected step some other way? Maybe use an outline icon style for other steps than the current one? This was probably not an issue before as it was clear which step a user is currently on.
>        image
> 
>     6. Step 5 has cut off content
>        image
> 
>     7. Step 7. Marking a dataset as deleted wraps the element
>        image
>        image
> 
>     8. On smaller screens, the vertical stepper takes up quite some space and it becomes tedious having to scroll up and down. This happens when trying to change the option for the same step or when trying to get to the next step every time. Maybe it makes sense to have `next`/`previous` buttons somewhere? Or have the stepper horizontally instead of vertically for smaller screens?
> 
> 
> (Screenshots from a width of 930px which is probably not that small after all) image image

@rekt-hard Thank you so much for reviewing the PR. Yes, this feature is indeed huge! I based the implementation on the initial XD examples, but as you mentioned, it makes sense now to go deeper into each step after all the changes. The detailed views for each step were something we agreed to tackle later, but now I see the need to address them.

@DererekMolnar, I think we should have a discussion tomorrow on how we want to proceed with these detailed views, as there haven’t been any suggestions so far. We can discuss it tomorrow for the initial plan for the next steps.

**Comment by ValentinFutterer** [comment]
>On smaller screens, the vertical stepper takes up quite some space and it becomes tedious having to scroll up and down. This happens when trying to change the option for the same step or when trying to get to the next step every time. Maybe it makes sense to have next/previous buttons somewhere? Or have the stepper horizontally instead of vertically for smaller screens?

The next button would be a great idea!

I have experimented today with a horizontal stepper, but i did not have much success, the stepper component doesnt offer a way to only show parts of it. So for e.g. i am at step 4 and i see 4, 5, 6, but i can navigate forwards or backwards with little arrows. Although there is a hacky way to maybe get this working, i havent tried this yet though. 

I have also experimented with making the stepper component itself smaller, and i am very confident that i can do that, this could save ~130px on space, depending on the exact values chosen.

There is also the option of making everything inside the components itself smaller - which would be a very tedious process, but effective. This work has to be done anyways to some extent, since the old css @media queries relied on there being more space for them, so now they trigger at unwanted times.

Lastly, i would suggest to trigger the auto-minimize of the sidebar way earlier, at around 1300px. I am aware this will effect the whole application, but i think its worth it.

**Comment by andresTabiTuwien** [comment]
I did the review and I found that in general everything is working well. I found some things that I think need to be checked, I'm not sure if they should be fixed in this PR or we should create another small issue to fix them.

Here is the list of things I found.

1. In step 1 in screens smaller than 1680 the end date of the calendar does not appear and in screens smaller than 1300 it appears behind the text of the project description.
step1
step1 1



2.  In step 3, the table and text area are cropped in screens smaller than 1000px when new data is tapped.
step3-new-data



3. In step 5, in the Storage tab, the fields and the text area are cut into screens smaller than 900px.
step5-830px



4.At step 8, on the Repository Search tab, the buttons for navigating the pagination are half visible.
step8



5. In step 10, if we have a list of costs and one of them is open, we will not be able to see the other cost items.
step10-costs
step10-costs-4

6.  In some steps we are getting this errror in console: "ERROR RuntimeError: NG0100: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'undefined'. Current value: '[object Object]'."  We can try to identify what generates the error and implement a strategy to avoid it. This coulb be as a new issue.
Steps that generate the error are 1, 2, 6, 8.
image

**Comment by lpandath** [comment]
Screenshot 2024-11-13 at 20 24 37

Thank you for the review, @andresTabiTuwien! 👍 

Here’s a summary of the changes made:
	1.	Step 1: Fixed display issues with the end date in the calendar. On screens smaller than 1680px, the end date now appears correctly, and for screens below 1300px, it no longer overlaps with the project description text.
	2.	Step 5 (Storage Tab): Adjusted layout so fields and the text area display correctly on screens smaller than 900px.
	3.	Step 8 (Repository Search Tab): Pagination navigation buttons are now fully visible.
	4.	Step 10: Fixed scrolling within the list of costs to ensure all items are accessible when one is expanded.
	5.	Entire Stepper: Added padding/margin adjustments to ensure step 11 is fully visible.

Dataset Table on Step 3: I couldn’t reproduce the issue fully, as the component needs a complete refactor. I tested it without the table, and it worked as expected based on the screenshot.

Console Error:  ExpressionChangedAfterItHasBeenCheckedError in steps 1, 2, 6, and 8 also resolved.

If everything looks good, I’ll go ahead squash the commits and merge.

**Comment by andresTabiTuwien** [comment]
Hi @lpandath, I have checked the points and they are all solved.  Everything works really well.

**Review by andresTabiTuwien — APPROVED** [review]
Everything works well. great job

**Comment by lpandath** [comment]
The PR has been merged. Due to internal conflicts, I had to merge it directly. I thoroughly tested it multiple times beforehand to ensure there are no issues.

**Pull request merged**